File: algorithms/18_ocaml/data_structures/dsu.ml
type dsu = {
  mutable parent : int array;
  rank : int array;
}

let make n = { parent = Array.init n (fun i -> i); rank = Array.make n 0 }

let rec find dsu x =
  if dsu.parent.(x) <> x then dsu.parent.(x) <- find dsu dsu.parent.(x);
  dsu.parent.(x)

let union dsu x y =
  let rx = find dsu x in
  let ry = find dsu y in
  if rx <> ry then begin
    if dsu.rank.(rx) < dsu.rank.(ry) then dsu.parent.(rx) <- ry
    else if dsu.rank.(rx) > dsu.rank.(ry) then dsu.parent.(ry) <- rx
    else begin
      dsu.parent.(ry) <- rx;
      dsu.rank.(rx) <- dsu.rank.(rx) + 1
    end
  end

let connected dsu x y = find dsu x = find dsu y

let () =
  let dsu = make 5 in
  union dsu 0 1;
  union dsu 1 2;
  union dsu 3 4;
  assert (connected dsu 0 2);
  assert (not (connected dsu 0 3));
  Printf.printf "[OCaml DSU] Disjoint set union verified\n"

File: algorithms/13_swift/data_structures/DSU.swift
final class DSU {
    private var parent: [Int]
    private var rank: [Int]

    init(_ n: Int) {
        parent = Array(0..<n)
        rank = Array(repeating: 0, count: n)
    }

    func find(_ x: Int) -> Int {
        if parent[x] != x { parent[x] = find(parent[x]) }
        return parent[x]
    }

    func union(_ x: Int, _ y: Int) {
        let rx = find(x), ry = find(y)
        if rx == ry { return }
        if rank[rx] < rank[ry] {
            parent[rx] = ry
        } else if rank[rx] > rank[ry] {
            parent[ry] = rx
        } else {
            parent[ry] = rx
            rank[rx] += 1
        }
    }

    func connected(_ x: Int, _ y: Int) -> Bool { find(x) == find(y) }
}

let dsu = DSU(5)
dsu.union(0, 1)
dsu.union(1, 2)
dsu.union(3, 4)
assert(dsu.connected(0, 2))
assert(!dsu.connected(0, 3))
print("[Swift DSU] Disjoint set union verified")
